compiler/z/codegen/OMRPeephole.cpp

```diff
@@ -448,6 +453,92 @@ OMR::Z::Peephole::attemptToReduceAGI(TR::Instruction* cursor)
    return performed;
    }
 
+bool
+OMR::Z::Peephole::attemptToReduceLToICM(TR::Instruction* cursor)
+   {
+   // Look for L/LTR instruction pair and reduce to LTR
+   TR::Instruction* prev = cursor->getPrev();
+   TR::Instruction* next = cursor->getNext();
+   if (next->getOpCodeValue() != TR::InstOpCode::LTR && next->getOpCodeValue() != TR::InstOpCode::CHI)
+      return false;
+
+   bool performed = false;
+   bool isICMOpportunity = false;
+
+   TR::S390RXInstruction* load= (TR::S390RXInstruction*) cursor;
+
+   TR::MemoryReference* mem = load->getMemoryReference();
+
+   if (mem == NULL) return false;
+   // We cannot reduce L to ICM if the L uses both index and base registers.
+   if (mem->getBaseRegister() != NULL && mem->getIndexRegister() != NULL) return false;
+
+   // L and LTR/CHI instructions must work on the same register.
+   if (load->getRegisterOperand(1) != next->getRegisterOperand(1)) return false;
+
+   if (next->getOpCodeValue() == TR::InstOpCode::LTR)
+      {
+      // We must check for sourceRegister on the LTR instruction, in case there was a dead load
+      //        L      GPRX, MEM       <-- this is a dead load, for whatever reason (e.g. GRA)
+      //        LTR    GPRX, GPRY
+      // it is wrong to transform the above sequence into
+      //        ICM    GPRX, MEM
+      if (next->getRegisterOperand(1) != ((TR::S390RRInstruction*)next)->getRegisterOperand(2)) return false;
+      isICMOpportunity = true;
+      }
+   else
+      {
+      // CHI Opportunty:
+      //      L  GPRX,..
+      //      CHI  GPRX, '0'
+      // reduce to:
+      //      ICM GPRX,...
+      TR::S390RIInstruction* chi = (TR::S390RIInstruction*) next;
+
+      // CHI must be comparing against '0' (i.e. NULLCHK) or else our
+      // condition codes will be wrong.
+      if (chi->getSourceImmediate() != 0) return false;
+
+      isICMOpportunity = true;
+      }
+
+   if (isICMOpportunity && performTransformation(comp(), "\nO^O S390 PEEPHOLE: Reducing L [%p] being reduced to ICM.\n", cursor))
+      {
+      // Prevent reuse of memory reference
+      TR::MemoryReference* memcp = generateS390MemoryReference(*load->getMemoryReference(), 0, cg());
```

Comment:
Use `mem` instead of `load->getMemoryReference()` ?

Reply:
Fixed in ece2f17